File: Cores/ExtremumScalaCore/src/main/scala/OSOL/Extremum/Core/Scala/Optimization/Nodes/TerminationViaMaxTime.scala
package OSOL.Extremum.Core.Scala.Optimization.Nodes

import OSOL.Extremum.Core.Scala.Optimization._
import OSOL.Extremum.Core.Scala.Optimization.Exceptions._

class TerminationViaMaxTime[Base, FuncType, V <: Optimizable[Base, FuncType]](override val nodeId: Int, val maxTime: Double)
  extends GeneralNode[Base, FuncType, V ](nodeId) {

  final private val parameterName = "startTime"

  final override def initialize(f: Map[String, FuncType] => FuncType, area: Area, state: State[Base, FuncType, V]): Unit = {
    try {
      val _ = state.getParameter[Long](parameterName)
      throw new ParameterAlreadyExistsException(parameterName)
    }
    catch {
      case _: NoSuchParameterException => state.setParameter(parameterName, System.nanoTime())
      case _ => throw new Exception("Unknown Exception")
    }
  }

  final override def process(f: Map[String, FuncType] => FuncType, area: Area, state: State[Base, FuncType, V]): Unit = { }

  final override def getCurrentCondition(f: Map[String, FuncType] => FuncType, area: Area, state: State[Base, FuncType, V]): Option[Int] = {
    if (1e-9 * (System.nanoTime() - state.getParameter[Long](parameterName)) > maxTime) Some(1)
    else Some(0)
  }

}

File: Cores/OSOL.Extremum.Core.DotNet/Optimization/Nodes/TerminationViaMaxTime.cs
using System;
using System.Collections.Generic;

namespace OSOL.Extremum.Core.DotNet.Optimization.Nodes
{
    public class TerminationViaMaxTime<TBase, TFuncType, TV>: GeneralNode<TBase, TFuncType, TV>
        where TV: class, IOptimizable<TBase, TFuncType>
    {
        private string ParameterName;
        private double MaxTime;

        public TerminationViaMaxTime(int nodeId, double maxTime, string parameterName = "startTime")
        {
            this.NodeId = nodeId;
            this.MaxTime = maxTime;
            this.ParameterName = parameterName;
        }

        public override void Initialize(Func<Dictionary<string, TFuncType>, TFuncType> f, Area area, State<TBase, TFuncType, TV> state)
        {
            int? value = null;
            try
            {
                value = state.GetParameter<int>(ParameterName);
            }
            catch (OptimizationExceptions.NoSuchParameterException e)
            {
                state.SetParameter<DateTime>(name: ParameterName, value: DateTime.Now);
            }
            finally
            {
                if(value.HasValue) throw new OptimizationExceptions.ParameterAlreadyExistsException(ParameterName);
            }
        }

        public override void Process(Func<Dictionary<string, TFuncType>, TFuncType> f, Area area, State<TBase, TFuncType, TV> state)
        {
            state.SetParameter(name: ParameterName, value: state.GetParameter<int>(ParameterName) + 1);
        }

        public override int? GetCurrentCondition(Func<Dictionary<string, TFuncType>, TFuncType> f, Area area, State<TBase, TFuncType, TV> state) => 
            (DateTime.Now - state.GetParameter<DateTime>(ParameterName)).TotalSeconds > MaxTime ? 1 : 0;

    }
}